Add towncrier for automatic changelog generation

wgreenberg opened the pull request:
blast from the past! resurrects https://github.com/certbot/certbot/pull/9803 with all of @bmw's changes. i figured instead of force-pushing a basically brand new branch and obliterating the old review, i'd just start from a clean slate

fixes #8272

bmw commented on newsfragments/8272.changed:
```diff
@@ -0,0 +1 @@
+Added towncrier for automatic changelog generation
```
i personally don't think our users need to hear about this and would recommend deleting it, but you can keep it if you want to

bmw commented on newsfragments/10342.fixed:
```diff
@@ -0,0 +1 @@
+Catches and ignores errors during the directory fetch for ARI checking so that these errors do not hinder the actual certificate issuance. Certbot also now uses the server value from the renewal configuration file for ARI checks instead of the server value from the current invocation of Certbot. This helps prevent ARI requests from going to the wrong server if the user changes CAs.
```
any reason not to keep these two changelog entries separate, deleting all but the first sentence here, letting it link to #10342, and creating a new 10339.fixed file for the

> Certbot now always uses the server value from the renewal configuration file for ARI checks instead of the server value from the current invocation of Certbot. This helps prevent ARI requests from going to the wrong server if the user changes CAs.

fix for #10339?

i also noticed this moved the first sentence here from "changed" to "fixed". if you meant to do this, that's fine with me, otherwise, i'd suggest renaming this file 10342.changed

bmw reviewed (changes requested):
i made a number of small suggestions, but this otherwise looks great to me!

thanks for picking this up again will

wgreenberg commented on certbot/CHANGELOG.md:
```diff
@@ -2,29 +2,6 @@
 
 Certbot adheres to [Semantic Versioning](https://semver.org/).
```
huh TIL about markdown comments. i kinda feel like this makes the changelog less legible in general, how would you feel instead about a line in the towncrier CI or release script which errors if the expected line is missing?

bmw replied:
we could do that. i think it a tradeoff between the complexity of our setup and the readability of our changelog. erring on the side of readability may indeed make sense

one other suggestion tho and if you don't like it, we can go with your idea

by default in markdown files, towncrier looks for the string `<!-- towncrier release notes start -->\n` according to https://towncrier.readthedocs.io/en/stable/configuration.html. i think part of my concern here is we're tying meaning to this sentence about semantic versioning that is not at all obvious just looking at the file

what if instead we followed towncrier's suggestion and used a comment so this file would look like

```suggestion
Certbot adheres to [Semantic Versioning](https://semver.org/).

<!-- towncrier will insert new entries here -->
```

and then we modify towncrier.toml to set start_string to `"<!-- towncrier will insert new entries here -->\n"`

this decouples the semantic versioning sentence and magic string for towncrier while also making it clear to people looking at changelog.md where new entries will go

do ya like this or do you still prefer your idea?

wgreenberg replied:
the default markdown start string is actually fairly terse and IMO wards off accidentally removing it:
```markdown
<!-- towncrier release notes start -->\n
```

bmw replied:
wanna use that then?

wgreenberg commented:
@ohemorange i suppose our guidance could be to make all newsfragments one-liners, since IMO the default multiline behavior is pretty wack, and disabling `all_bullets` (afaict) would require everyone to include a `* ` prefix to their fragment

ohemorange commented on newsfragments/10259.fixed:
```diff
@@ -0,0 +1 @@
+Previously, we claimed to set FAILED_DOMAINS and RENEWED_DOMAINS env variables for use by post-hooks when certificate renewals fail, but we were not actually setting them. Now, we are.
```
down to enable wrap, but then the fragments should be a single line because the double wrap looks bad

ohemorange commented:
I agree the default multiline behavior is a little silly, but requiring all oneliners also seems bad. how would you do #10378 as a one-liner?

wgreenberg commented:
```markdown
* Removed `acme.challenges.TLSALPN01Response`, `acme.challenges.TLSALPN01`, `acme.standalone.TLSServer`, and `acme.standalone.TLSALPN01Server`. Also removed parameter `alpn_protocols` from `acme.crypto_util.probe_sni`
```

it's not good!!

wgreenberg commented:
so i guess `wrap=true` still copies the fragment's original whitespace, so we actually can't combine it with manual wrapping. but with that in mind, the default multiline functionality is probably fine:

fragment:
```
Removed several deprecated classes
- Removed `acme.challenges.TLSALPN01Response`
- Removed `acme.challenges.TLSALPN01`
- Removed parameter `alpn_protocols` from `acme.crypto_util.probe_sni`
- Removed `acme.standalone.TLSServer`
- Removed `acme.standalone.TLSALPN01Server`
```

changelog result:
```markdown
### Changed

- Removed several deprecated classes
  - Removed `acme.challenges.TLSALPN01Response`
  - Removed `acme.challenges.TLSALPN01`
  - Removed parameter `alpn_protocols` from `acme.crypto_util.probe_sni`
  - Removed `acme.standalone.TLSServer`
  - Removed `acme.standalone.TLSALPN01Server`
```
```

ohemorange commented:
Multiline with long lines now looks like this:

<img width="764" height="498" alt="Screenshot 2025-07-30 at 5 39 40 PM" src="https://github.com/user-attachments/assets/19bfbdc9-1a4d-4f73-901a-ea988d133f23" />

I agree this is solid.

ohemorange commented:
Not sure if you sure my comment at https://github.com/certbot/certbot/pull/10379#discussion_r2243784269, so bumping it here

wgreenberg commented on towncrier.toml:
```diff
@@ -0,0 +1,24 @@
+filename = "certbot/CHANGELOG.md"
+title_format = "## {version} - {project_date}"
+start_string = "Certbot adheres to [Semantic Versioning](https://semver.org/).\n"
+issue_format = "[#{issue}](https://github.com/certbot/certbot/issues/{issue})"
```
right, fragment names can either be issue numbers (matched w/ this regex) or so-called "orphan" fragments (which start with a +). issue numbers are formatted w/ the github issue link, and orphans aren't

ohemorange replied:
got it, thanks! sounds like the check for valid filenames is done elsewhere, and this is just for matching numbers to create issue links. I assumed this was also the check for valid filenames.

ohemorange commented:
slight change to add wrapping in case @bmw wants to take another look
